File: 프로그래머스/lv2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.java
import java.util.*;
class Solution {
    private static int result = 0;
    public int solution(int[] queue1, int[] queue2) {
        long total1 = 0;
        long total2 = 0;
        Deque<Integer> list1 = new LinkedList<>();
        Deque<Integer> list2 = new LinkedList<>();
        
        for (int i = 0; i < queue1.length; i++) {
            list1.add(queue1[i]);
            list2.add(queue2[i]);
            total1 += queue1[i];
            total2 += queue2[i];
        }
        
        long TSum = total1 + total2;
        long half = TSum/2;
        if(half < Arrays.stream(queue2).max().getAsInt() || half < Arrays.stream(queue1).max().getAsInt()|| TSum %2 != 0){
            return -1;
        }  // 애시당초 안되는 애들은 제외
        
        dfs(list1,list2,half);
        return result;
    }
    private static void dfs(Deque<Integer> list1, Deque<Integer> list2, long half) {
        // 차피 한 쪽만 맞춰지면 나머지는 신경안써도 된다.
        long listSum = 0;
        for (int a : list1) {
            listSum += a;
        }
        int temp = 0;
        int count = 0;

        while (list2.size() > 0) {
            if (listSum > half) {
                temp = list1.poll();
                listSum -= temp;
                count += 1;
            } else if (listSum == half) {
                result = count;
                return;
            } else {
                temp = list2.poll();
                listSum += temp;
                list1.add(temp);
                count += 1;
            }
        }
        
        result = -1;
    }
}

File: 프로그래머스/lv2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.py
from collections import deque

def solution(queue1, queue2):
    answer = -1
    N = len(queue1)
    total1, total2 = sum(queue1), sum(queue2)
    origin = queue2[:]
    
    queue1, queue2 = deque(queue1), deque(queue2)
    target = total1 + total2
    # 만약 두개의 큐의 합이 홀수라면 합을 같게 만들수 없기 때문에 바로 리턴한다
    if target % 2 == 1:
        return answer
    
    target //= 2
    count = 0
    # queue1을 기준으로 target 값보다 크면 pop 아니면 push를 한다
    # queue1이 queue2랑 같거나 or queue2에서 더이상 pop할 원소가 없을때 탐색을 멈춘다
    while queue2:
        if total1 == target:
            answer = count
            break
        # 1. queue1의 총합이 target 값보다 작을 경우
        if total1 < target:
            n = queue2.popleft()
            total1 += n
            queue1.append(n)
        # 2. queue2의 총합이 target 값보다 클 경우
        else:
            total1 -= queue1.popleft()
        count += 1

        
    return answer